platform/darwin/src/MGLOfflineStorage.mm

```diff
@@ -264,6 +264,62 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
     }
 }
 
+#pragma mark Offline merge methods
+
+- (void)addContentsOfFile:(NSString *)filePath withCompletionHandler:(MGLBatchedOfflinePackAdditionCompletionHandler)completion {
+    NSFileManager *fileManager = [NSFileManager defaultManager];
+    if (![fileManager isWritableFileAtPath:filePath]) {
+        [NSException raise:NSInvalidArgumentException format:@"The file path must be writable"];
+    }
+    
+    __weak MGLOfflineStorage *weakSelf = self;
+    [self _addContentsOfFile:filePath withCompletionHandler:^(NSError * _Nullable error) {
+        if (error && completion) {
+            completion(error);
+        } else {
+            MGLOfflineStorage *strongSelf = weakSelf;
+            [strongSelf getPacksWithCompletionHandler:^(NSArray<MGLOfflinePack *> *packs, NSError * _Nullable error) {
+                for (MGLOfflinePack *pack in strongSelf.packs) {
+                    [pack invalidate];
+                }
+                strongSelf.packs = [packs mutableCopy];
+                if (completion) {
+                    completion(error);
+                }
+            }];
+        }
+        
+    }];
+}
+
+- (void)addContentsOfURL:(NSURL *)fileURL withCompletionHandler:(MGLBatchedOfflinePackAdditionCompletionHandler)completion {
+    [self addContentsOfFile:fileURL.absoluteString withCompletionHandler:completion];
+}
+
+- (void)_addContentsOfFile:(NSString *)filePath withCompletionHandler:(MGLBatchedOfflinePackAdditionCompletionHandler)completion {
+    self.mbglFileSource->mergeOfflineRegions(std::string(static_cast<const char *>([filePath UTF8String])), [&, completion](mbgl::expected<mbgl::OfflineRegions, std::exception_ptr> result) {
+        NSError *error;
+        NSMutableArray *packs;
+        if (!result) {
+            error = [NSError errorWithDomain:MGLErrorDomain code:-1 userInfo:@{
+                                                                               NSLocalizedDescriptionKey: @(mbgl::util::toString(result.error()).c_str()),
```

Comment:
The error strings coming from mbgl are fairly low-level and specific, such as:

https://github.com/mapbox/mapbox-gl-native/blob/6e171a62a9cc07ddbb0765e3021e46ebd60313df/platform/default/mbgl/storage/offline_database.cpp#L661

Since a localized description is meant for display to the user, it should be something higher-level like “Unable to add offline packs from the file at…”. (Remember to make this string localizable.) The error string from mbgl should go in `NSLocalizedFailureReasonKey`.

Reply:
Most of the errors coming from `offline_database` are low level I may be wrong but I think that we rely on the developer to display the proper error message after receiving an `NSError` object.

Reply:
That may be, but the convention is still for the localized description to be as user-friendly as possible, including localization, and for the failure reason to contain a more technical explanation. See `MGLMapViewImpl::onDidFailLoadingMap()` for inspiration.